src/DotNetty.Common/Utilities/HashedWheelTimer.cs

```diff
@@ -0,0 +1,712 @@
+            while (this.StartTime == PreciseTimeSpan.Zero)
+            {
+                this.startTimeInitialized.Wait(this.CancellationToken);
+            }
+        }
+
+        public ISet<ITimeout> Stop()
+        {
+            if (XThread.CurrentThread == this.workerThread)
+            {
+                throw new InvalidOperationException($"{nameof(HashedWheelTimer)}.stop() cannot be called from timer task.");
+            }
+
+            if (Interlocked.CompareExchange(ref this.workerStateVolatile, WorkerStateShutdown, WorkerStateStarted) != WorkerStateStarted)
+            {
+                // workerState can be 0 or 2 at this moment - let it always be 2.
+                if (Interlocked.Exchange(ref this.workerStateVolatile, WorkerStateShutdown) != WorkerStateShutdown)
+                {
+                    Interlocked.Decrement(ref instanceCounter);
+                }
+
+                return new HashSet<ITimeout>();
+            }
+
+            try
+            {
+                this.cancellationTokenSource.Cancel();
+                this.workerThread.Join(100);
+            }
+            finally
+            {
+                Interlocked.Decrement(ref instanceCounter);
+            }
+            return this.worker.UnprocessedTimeouts();
+        }
+
+        public ITimeout NewTimeout(ITimerTask task, TimeSpan delay)
+        {
+            if (task == null)
+            {
+                throw new ArgumentNullException(nameof(task));
+            }
+            if (this.ShouldLimitTimeouts())
+            {
+                long pendingTimeoutsCount = Interlocked.Increment(ref this.PendingTimeouts);
+                if (pendingTimeoutsCount > this.maxPendingTimeouts)
+                {
+                    Interlocked.Decrement(ref this.PendingTimeouts);
+                    throw new RejectedExecutionException($"Number of pending timeouts ({pendingTimeoutsCount}) is greater than or equal to maximum allowed pending timeouts ({this.maxPendingTimeouts})");
+                }
+            }
+
+            this.Start();
+
+            // Add the timeout to the timeout queue which will be processed on the next tick.
+            // During processing all the queued HashedWheelTimeouts will be added to the correct HashedWheelBucket.
+            TimeSpan deadline = CeilTimeSpanToMilliseconds((PreciseTimeSpan.Deadline(delay) - this.StartTime).ToTimeSpan());
+            var timeout = new HashedWheelTimeout(this, task, deadline);
+            this.timeouts.TryEnqueue(timeout);
+            return timeout;
+        }
+
+        bool ShouldLimitTimeouts() => this.maxPendingTimeouts > 0;
+
+        static void ReportTooManyInstances() =>
+            Logger.Error($"You are creating too many {nameof(HashedWheelTimer)} instances. {nameof(HashedWheelTimer)} is a shared resource that must be reused across the process,so that only a few instances are created.");
+
+        static TimeSpan CeilTimeSpanToMilliseconds(TimeSpan time)
```

Comment:
yes

---
In reply to: [114475437](https://github.com/Azure/DotNetty/pull/242#discussion_r114475437) [](ancestors = 114475437)